bindings/python/mappings.py

```diff
@@ -32,81 +33,462 @@ class InconsistentTriplesError(MappingError):
+        elif p == returns:
+            Dreturns[s].append(o)
+        elif p == first:
+            Dfirst[s].append(o)
+        elif p == rest:
+            Drest[s].append(o)
+
+    d = defaultdict(list)
+    for func, lst in Dreturns.items():
+        input_iris = []
+        for exp in Dexpects.get(func, ()):
+            if exp in Dfirst:
+                while exp in Dfirst:
+                    input_iris.append(Dfirst[exp])
+                    if exp not in Drest:
+                        break
+                    exp = Drest[exp]
+            else:
+                # Support also misuse of FNO, where fno:expects refers
+                # directly to input individuals
+                input_iris.append(exp)
+
+        for ret in lst:
+            if ret in Dfirst:
+                while ret in Dfirst:
+                    d[Dfirst[ret]].append((func, input_iris))
+                    if ret not in Drest:
+                        break
+                    ret = Drest[ret]
             else:
-                sources.append(input)
-        return sources
+                # Support also misuse of FNO, where fno:returns refers
+                # directly to the returned individual
+                d[ret].append((func, input_iris))
+
+    return d
+
+
+def mapping_route(
+        target, sources, triples,
+        function_repo=None,
+        function_mappers=(fno_mapper, ),
+        default_costs={'function': 10.0, 'mapsTo': 2.0, 'instanceOf': 1.0,
+                       'subClassOf': 1.0, 'value': 0.0},
+        mapsTo='http://emmo.info/domain-mappings#mapsTo',
+        instanceOf='http://emmo.info/datamodel#instanceOf',
+        subClassOf='http://www.w3.org/2000/01/rdf-schema#subClassOf',
+        #type='http://www.w3.org/1999/02/22-rdf-syntax-ns#type',
+        #description='http://purl.org/dc/terms/description',
+        hasName='http://www.w3.org/2000/01/rdf-schema#label',
+        hasUnit='http://emmo.info/datamodel#hasUnit',
+        hasCost=':hasCost',
+        hasValue='http://emmo.info/datamodel#hasValue',
+):
+    """Find routes of mappings from any source in `sources` to `target`.
+
+    This implementation supports transitivity, subclasses.
+
+    Arguments:
+        target: IRI of the target in `triples`.
+        sources: Sequence of source IRIs in `triples`.
+        triples: Sequence of (subject, steptype, object) triples.
+            It is safe to pass a generator expression too.
+        mapsTo: How 'mapsTo' is written in `triples`.
+        subClassOf: How 'subClassOf' is written in `triples`.  Set it
+            to None if subclasses should not be considered.
+        function_mappers: Sequence of mapping functions that takes `triples`
+            as argument and return a dict mapping output IRIs to a list
+            of `(function_iri, [input_iris, ...])` tuples.
+        hasName: How 'hasName' is written in `triples`.  Used for naming
+            function in put parameters.  The default is to use rdfs:label.
```

Comment:
Does it check all rdfs:label? And also, can we ask for rdfs:label in one ontology and skos:prefLabel in another?

Reply:
The argument has been updated to `label`. It checks all rdfs:label in the ontology.

It does not support asking for rdfs:label in one ontology and skos:prefLabel in another, hence mapping functions must be labeled consistently.